wordpress-mobile/WordPress-iOS
# App Extensions: Added categories module

![categories](https://user-images.githubusercontent.com/154014/36814785-4193a680-1c9e-11e8-8c15-cb50c70e9b3e.gif)

## Discussion
This PR adds the categories module to the new app extensions. 

### Picker Activation
The category module cell will not launch the category picker if a site only contains a single category (the disclosure chevron will be hidden as well):

![no_select](https://user-images.githubusercontent.com/154014/36814783-3ce13e68-1c9e-11e8-94a1-348aec365a44.jpg)

### Defaults
**_IF_** we are able to retrieve the default category for a given site, it will be pre-selected in the category picker screen and shown on the right side of the category module cell:

![default_cat](https://user-images.githubusercontent.com/154014/36813520-7e4b4b68-1c9a-11e8-959c-68db973e7331.jpg)

However, if for some reason we are not able determine the default category (customer is not VPNed/proxied, no permissions to access site settings, etc) then the word 'Default' will be displayed:

![no_default_cat](https://user-images.githubusercontent.com/154014/36813609-c103d5ce-1c9a-11e8-8787-f3a9cca65ce0.jpg)

Note that the net effect of leaving this default value and saving/publishing the post is the same — a post is created using the default category.

### Data
The Extensions Core Data model (that lives in the shared container) is updated via lightweight migration. A new, optional `String` field was added: `postCategories` to `PostUploadOperation`.

### Network
Two new network functions were added to `AppExtensionsService` 
* `fetchSettingsForSite(_:onSuccess:onFailure:)` — used to get the default category
* `fetchCategoriesForSite(_:onSuccess:onFailure:)` — fetches all categories for a site

In order to prevent expensive network calls from occurring every time a user taps a new site, category info for each site is temporarily stashed in a `NSCache` located in `ShareExtensionAbstractViewController.swift`. This cache is purged when the customer either pulls-to-refresh on the site picker screen, or when exiting the app extension.

### Analytics
Two new analytic values were added and are being tracked: `wpios_*_extension_categories_opened` and `wpios_*_extension_categories_selected`.

Fixes #8142 

## Testing
Make sure to test both the "Save as Draft" action extension as well as the share extension. Both extensions use the same code, but you can never be too careful. 😄 

Testing ideas:
* Test sharing images as well as web/text. Verify the categories are saved properly.
* Try setting some categories and then switching sites (the categories will clear). Verify the correct categories made it to the server.
* Verify the category tree hierarchy — make sure it matches WPiOS (and the server)
* Make sure the cache is working properly and loading categories for each site. If you pull-to-refresh the sites list or cancel, categories should reload when a new site is selected.
* Test the category screens with a spotty network connection.

@frosty and @kwonye would you mind taking a 👀 ?

/cc @folletto for design
